pkg/startup/v15/startup.go

```diff
@@ -251,3 +254,55 @@ func (s *startup) writeFiles(role api.AgentPoolProfileRole, w writers.Writer, ho
 
 	return w.Close()
 }
+
+// WriteSearchDomain queries for the search domain and writes to
+// /etc/dhcp/dhclient-eth0.conf.  This is for private api clusters only
+func (s *startup) WriteSearchDomain(ctx context.Context, log *logrus.Entry) error {
+	spp := &s.cs.Properties.WorkerServicePrincipalProfile
+
+	s.log.Info("creating clients")
+	authorizer, err := azureclient.NewAuthorizer(spp.ClientID, spp.Secret, s.cs.Properties.AzProfile.TenantID, "")
+	if err != nil {
+		return err
+	}
+
+	ncli := network.NewInterfacesClient(ctx, log, s.cs.Properties.AzProfile.SubscriptionID, authorizer)
+	var dnsString string
+	nic, err := ncli.GetVirtualMachineScaleSetNetworkInterface(ctx, s.cs.Properties.AzProfile.ResourceGroup, "ss-master", "0", "nic", "")
+	if err != nil {
+		return err
+	}
+	if nic.DNSSettings != nil {
+		dnsString = to.String(nic.DNSSettings.InternalDomainNameSuffix)
+	}
+	s.log.Info("writing custom dns")
+
+	domainSettings := []string{
+		"interface \"eth0\" {\n",
+		fmt.Sprintf("    supersede domain-name \"%s\";\n", dnsString),
+		fmt.Sprintf("    supersede domain-search \"%s\";\n", dnsString),
+		"}\n",
+	}
+
+	// write to dnsmasq.conf and see if that is sufficient
+	err = writeContentsToFile("/host/etc/dhcp/dhclient-eth0.conf", domainSettings)
+	if err != nil {
+		return err
+	}
+
+	err = writeContentsToFile("/host/etc/dnsmasq.conf", []string{fmt.Sprintf("server=/%s/%s\n", dnsString, constants.AzureNameserver)})
+	return err
+}
+
+func writeContentsToFile(filename string, contents []string) error {
```

Comment:
use ioutil.WriteFile

Reply:
From looking at the description of `ioutil.WriteFile`, I did not see an append mode.  For the dnsmasq.conf I needed to append the `server=` string.  Is this possible with `ioutil.WriteFile`?  The many examples that I looked at showed using the `os.OpenFile` method with the `os.O_APPEND|os.O_CREATE|os.O_WRONLY`.

Reply:
What's wrong with opening the file using `os.O_APPEND`? We only use this helper twice, could just call `Write` twice against the open file?

Reply:
On one of the files we want to create and on one of the files we want to append.  Having this code two times in the same function seemed to violate DRY so I created a helper function on the side that accommodates both calls.

Reply:
Ooh, sorry, didn't see the O_APPEND, ignore me